### PR TITLE
feat(streams): limit usage in body mixin #2164

### LINKED ISSUE #2164: fetch: limit web streams usage in body mixin methods
The fetch spec requires us to create a new web Readable stream for every `Response` and `Request` object, which can likely be avoided in most cases.

Given the following examples:
```js
new Response('abc')
new Response(new TextEncoder().encode('abc'))
new Response(new Blob(['abc']))
// etc.
```

A web stream is created and then converted *back* to a string! The following diff improves `.text()` performance when the body passed is a string by **10%**, as a simple demo (we could further limit web stream usage for more performance gains!!!).

```diff
diff --git a/lib/fetch/response.js b/lib/fetch/response.js
index 1029dbef..c2d2bd3d 100644
--- a/lib/fetch/response.js
+++ b/lib/fetch/response.js
@@ -154,7 +154,7 @@ class Response {
     // 4. If body is non-null, then set bodyWithType to the result of extracting body.
     if (body != null) {
       const [extractedBody, type] = extractBody(body)
-      bodyWithType = { body: extractedBody, type }
+      bodyWithType = { body: extractedBody, type, source: body }
     }

     // 5. Perform initialize a response given this, init, and bodyWithType.
diff --git a/lib/fetch/util.js b/lib/fetch/util.js
index 400687ba..9e4302a9 100644
--- a/lib/fetch/util.js
+++ b/lib/fetch/util.js
@@ -824,6 +824,11 @@ function fullyReadBody (body, processBody, processBodyError) {
   //    with taskDestination.
   const errorSteps = (error) => queueMicrotask(() => processBodyError(error))

+  if (typeof body.source === 'string') {
+    successSteps(new TextEncoder().encode(body.source))
+    return
+  }
+
   // 4. Let reader be the result of getting a reader for body’s stream.
   //    If that threw an exception, then run errorSteps with that
   //    exception and return.
```

cc @anonrig

### PR DESCRIPTION
**Pull request by rozzilla**
## This relates to...
Fixes #2164 

## Rationale
When a `Response` or `Request` body is created from a `string` or `Uint8Array`, the body mixin methods (text(), json(), arrayBuffer(), blob(), bytes()) currently pipe the source through a `ReadableStream` even though the bytes are already available in `body.source`. This adds unnecessary overhead from stream construction, pump microtasks, and chunk collection.

## Changes
1. In `consumeBody()`, before calling `fullyReadBody()`, check if `body.source` is a `string` or `Uint8Array`. If so, bypass the stream round-trip and pass the source bytes directly to `successSteps`
2. Add tests covering `string` sources (all 5 body mixin methods, Request, multibyte, empty), `Uint8Array` sources (text, arrayBuffer, bytes, blob), body unusability after consumption (bodyUsed, double-read rejection), and non-shortcuttable sources (ReadableStream, Blob, null) to ensure the normal path is unaffected.

### Features
N/A

### Bug Fixes
N/A

### Breaking Changes and Deprecations
N/A

## Status

- [x] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [x] Tested
- [x] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md#developers-certificate-of-origin

### CODE REVIEW

**Review by KhafraDev — CHANGES_REQUESTED** [review]
The issue with optimizations like these (if it's even more performant, I'm not sure that it is) is that it encourages people to only use string or Uint8Array bodies to squeeze more performance out of node. I don't want to recreate CrankshaftScript.

**Comment by mcollina** [comment]
How much is more performant? I think that in most cases, those optimizations are good for Node.js.

**Comment by KhafraDev** [comment]
It wouldn't matter. The implementation is incorrect as it doesn't perform the same validation readAllBytes does (ensuring the result of reader.read() is a Uint8Array). After that I'm not sure, 1 less promise, for the extra maintenance burden and guaranteed bugs there will be?
